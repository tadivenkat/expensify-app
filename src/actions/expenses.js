import database from '../firebase/firebase';

export const addExpenseAction = (expense) => {
    return {
        type: 'ADD_EXPENSE',
        payload: expense
    };
};

export const startAddExpenseAction = (expenseData = {}) => {
    const {description = '', amount = 0, note = '', createdAt = 0} = expenseData;
    const expense = {description, amount, note, createdAt};
    return (dispatch) => {
        return database.ref('expenses').push(expense).then((data) => {
            dispatch(addExpenseAction({
                id: data.key,
                ...expense
            }));
        }).catch((error) => {
            console.log("Error adding expense", error);
        });
    }
};

export const removeExpenseAction = ({id} = {}) => {
    return {
        type: 'REMOVE_EXPENSE',
        id 
    }
};

export const editExpenseAction = ({id, amount, description, note, createdAt} = {}) => {
    return {
        type: 'EDIT_EXPENSE',
        payload: {
            id,
            amount,
            description,
            note,
            createdAt
        }
    }
};

export const setExpensesAction = (expenses = []) => {
    return {
        type: 'SET_EXPENSES',
        payload: expenses
    }
};

export const startSetExpensesAction = () => {
    const expenses = [];
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpensesAction(expenses));
        }).catch((error) => {
            console.log("Error retreiving expenses", error);
        });
    };
};