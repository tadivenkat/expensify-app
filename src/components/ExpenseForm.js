import React from 'react';
import moment from 'moment';
import DatePicker from 'react-datepicker';


class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            id: props.expense ? props.expense.id : '',
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? props.expense.amount / 100 : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt).valueOf() : moment().valueOf(),
            error: ''
        };
    };
    
    onDescriptionChange = (event) => {
        const description = event.target.value;
        this.setState((currentState) => {
            return {
                description
            };
        });
    };

    onNoteChange = (event) => {
        const note = event.target.value;
        this.setState((currentState) => {
            return {
                note
            };
        });
    };

    onAmountChange = (event) => {
        const amount = event.target.value;
        if (!amount || amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState((currentState) => {
                return {
                    amount
                };
            });
        }
    }

    onDateChange = (date) => {
        this.setState((currentState) => {
            return {
                createdAt: moment(date).valueOf()
            };
        });
    }

    onFormSubmit = (event) => {
        event.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => {
               return {
                   error: 'Please enter description and amount' 
               };
            });
        } else {
            this.setState(() => {
                return {
                    error: '' 
                };
            });
            // Construct expense object and set it on props to make it accessible to the parent component
            this.props.onSubmit({
                id: this.state.id,
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt
            });
        }
    }
    render() {
        return (
            <form className="form" onSubmit={this.onFormSubmit}>
                {this.state.error && <p className="bg-danger text-white">{this.state.error}</p>}
                <div className="form-group">
                    <label for="description">Description</label>
                    <input 
                        type="text"
                        id="description"
                        className="form-control"
                        placeholder="Enter Description" 
                        autoFocus 
                        value={this.state.description} 
                        onChange={this.onDescriptionChange} />
                </div>
                <div className="form-group">
                    <label for="amount">Amount</label>
                    <input
                        id="amount"
                        type="number"
                        className="form-control" 
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}/>
                </div>
                <div className="form-group">
                    <label for="note">Note</label>
                    <textarea 
                        id="note"
                        name="note"
                        className="form-control" 
                        placeholder="Add a Note for this Expense"
                        rows="7"
                        value={this.state.note}
                        onChange={this.onNoteChange}>
                    </textarea>
                </div>
                <div className="form-group">
                    <label for="createdAt">Created At</label>
                    <DatePicker
                        id="createdAt"
                        name="createdAt"
                        className="form-control" 
                        todayButton={"Today"}
                        selected={new Date(this.state.createdAt)} 
                        onChange={this.onDateChange}
                        title="Created At"
                        dateFormat="dd-MMM-yyyy"/>
                </div>
                <div>
                    <button onClick={this.onFormSubmit} className="btn btn-primary">Save Expense</button>
                </div>                    
            </form>                                        
        );
    }
}

export default ExpenseForm;