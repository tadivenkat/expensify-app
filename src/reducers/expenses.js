
export default (state = [], action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.payload];
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.payload.id) {
                    return {...expense, ...action.payload};
                } else {
                    return expense;
                }
            });
        default: 
            return state;
    }
};