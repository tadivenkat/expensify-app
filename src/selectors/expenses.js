// Return subset of expenses applying the filters.

export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        let startDateMatch = false;
        let endDateMatch = false;
        const textMatch = expense.description.toLowerCase().indexOf(text.toLowerCase()) >= 0;
        if (!!startDate) {
            if (!!expense.createdAt) {
                startDateMatch = startDate <= expense.createdAt;
            } else {
                startDateMatch = true;
            }
        } else {
            startDateMatch = true;
        }
        if (!!endDate) {
            if (!!expense.createdAt) {
                endDateMatch = endDate >= expense.createdAt;
            } else {
                endDateMatch = true;
            }
        } else {
            endDateMatch = true;
        }
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expense1, expense2) => {
        switch (sortBy) {
            case 'amount':
                return expense1.amount - expense2.amount;
            case 'date':
                return expense1.createdAt - expense2.createdAt;
            default:
                return expense1.amount - expense2.amount;
        }
    });
};