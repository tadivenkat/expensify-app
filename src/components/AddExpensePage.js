import React from 'react';
import {connect} from 'react-redux';
import {startAddExpenseAction} from '../actions/expenses';
import ExpenseForm from './ExpenseForm';
import Header from './Header';

export class AddExpensePage extends React.Component {
        onSubmit = (expense) => {
        this.props.onSubmit(expense);
        this.props.history.push('/dashboard');            
    }    
    render() {
        return (
            <div>
                <Header/>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit}/>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        onSubmit: (expense) => dispatch(startAddExpenseAction(expense))
    };
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage);