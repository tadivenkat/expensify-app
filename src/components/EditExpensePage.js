import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpenseAction, removeExpenseAction} from '../actions/expenses';

const EditExpensePage = (props) => {
    return (
        <div>
            <h1>Editing Expense: {props.match.params.id}</h1>
            <ExpenseForm 
                expense={props.expense}
                onSubmit={(expense) => {
                    console.log(expense);
                    props.dispatch(editExpenseAction(expense));
                    props.history.push('/');
                }}
            />
            <button onClick={() => {
                props.dispatch(removeExpenseAction(props.expense));
                props.history.push("/");
            }}>Remove</button>
        </div>
    );
};

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

export default connect(mapStateToProps)(EditExpensePage);