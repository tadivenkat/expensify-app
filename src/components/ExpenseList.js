import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import filterExpenses from '../selectors/expenses';
import {totalExpenditure} from '../selectors/expenses';
import moment from 'moment';
import numeral from 'numeral';

export const ExpenseList = (props) => {
    const {text, sortBy, startDate, endDate} = props.filters;
    return (
        <div>
            <h1>Expense List</h1>
            <p>
                You have {props.expenses.length} {props.expenses.length === 1 ? 'expense' : 'expenses'} with a total 
                expenditure of {numeral(totalExpenditure(props.expenses)/100).format('$ 0,0.00')}
            </p>
            {props.expenses.map((expense) => <ExpenseListItem 
                                                    key={expense.id} 
                                                    expense={expense}/>)}
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        expenses: filterExpenses(state.expenses, state.filters),
        filters: state.filters
    };
};

export default connect(mapStateToProps)(ExpenseList);