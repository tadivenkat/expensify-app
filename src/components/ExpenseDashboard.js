import React from 'react';
import ExpenseList from './ExpenseList';
import Filters from './Filters';
import Header from './Header';

class ExpenseDashboard extends React.Component {
  render() {
    return (
      <div>
        <Header/>
        <Filters/>
        <ExpenseList/>
      </div>
    );
  }
}

export default ExpenseDashboard;
