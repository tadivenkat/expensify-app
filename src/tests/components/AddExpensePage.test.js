import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixtures/expenses';

let onSubmit, history, wrapper;

beforeEach(() => {
    onSubmit = jest.fn();
    history = {push: jest.fn()};
    wrapper = shallow(<AddExpensePage onSubmit={onSubmit} history={history}/>);
});

test('Should render AddExpensePage correctly', () => {
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('Should handle onSubmit event', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith('/dashboard');
    expect(onSubmit).toHaveBeenLastCalledWith(expenses[2]);
});