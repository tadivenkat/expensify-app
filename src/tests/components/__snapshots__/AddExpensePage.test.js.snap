// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render AddExpensePage correctly 1`] = `
<div>
  <Connect(Header) />
  <h1>
    Add Expense
  </h1>
  <div
    className="content-container"
  >
    <ExpenseForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
